src/SegmentingContainer.php

```diff
@@ -0,0 +1,110 @@
+<?php
+
+namespace Dhii\Container;
+
+use Psr\Container\ContainerInterface;
+use function array_filter;
+use function ltrim;
+
+/**
+ * A container implementation that wraps around another to allow the container to be segmented into sub-containers.
+ *
+ * Segmenting containers are intended to be used with hierarchical keys, i.e. keys that use a delimiter to indicate
+ * segments in the key, such as "some/test/key" or "some.deep.config.value".
+ *
+ * A segmenting container can yield 2 different kinds of results:
+ * * values
+ * * "segments"
+ *
+ * When `get()` is called, the key is split into segments according to the delimiter, and those segments are appended
+ * to an internal segment list (more on this further below) to obtain the full key.
+ *
+ * If the full key corresponds to a value in the inner container, the value for that key is returned.
+ *
+ * Otherwise, a new segmenting container is created and returned. This segmenting container will store the full key
+ * that resulted in its creation as an internal segment list, which will be automatically prepended to all keys that
+ * are requested through its `get()` method.
+ *
+ * Example usage:
+ *      Consider the hierarchy:
+ *          [
+ *              "config" => [
+ *                  "db" => [
+ *                      "host" => "localhost",
+ *                      "port" => 3306
+ *                  ]
+ *              ]
+ *          ]
+ *
+ *      The segmenting container can create a container that directly provides the "host" and "port":
+ *          $config = new SegmentingContainer($c, '.');
+ *          $dbConfig = $config->get('config.db');
+ *          $dbConfig->get("host"); // "localhost"
+ *          $dbConfig->get("port"); // 3306
+ *
+ * @since [*next-version*]
+ */
+class SegmentingContainer implements ContainerInterface
```

Comment:
I absolutely love this decorator! And the description is pretty good. But after not seeing this class for a long time, I had forgotten how it works, and couldn't immediately understand it _correctly_ from just the description. The confusing part was that this class actually only works on the key, and the internal structure is irrelevant, i.e. it is not required to be a hierarchy of containers: a flat map would work just fine, e.g.

```
[
	'config.db.host' => 'localhost',
	'config.db.port' => 3306,
]
```

However, the example in the docs suggests something else. So, I would say that what this decorator does is that it allows working with a flat map where keys are segmented as if it was a hierarchy. I.e. with the example values above, you could get `'localhost'` by doing `$c->get('config')->get('db')->get('host')`.

What do you think about adding this disambiguation to the class docs, @mecha?

Reply:
Do we have a container that does the opposite? As in, if you had a _container_ hierarchy (like in your original doc), and you'd have to do `$c->get('config')->get('db')->get('host')` to get the host, it would allow accessing that value by doing `$c->get('config.db.host')`. I thought there was something like that.

Reply:
I need one like this for objects/arrays, so as to avoid making unnecessary assumptions about their structure. Doing `isset($obj->key1->key2->key3)` is very error prone. I would like to add one such implementation that deals with objects/arrays so that `key3` can simply be checked by `$c->has('key1/key2/key3')` etc.

Reply:
So, #7 adds the ability to transform a data structure into a container hierarchy. Now is the time to add this container. `PathContainer` seems like a fitting name. Not sure if belongs to this PR, or #7, or a completely new one. I'd say, let's just do what's necessary to get this one merged, and then we see.

Reply:
Agreed. I'll update the class doc; not sure how I managed to get the example usage wrong.

Re. `PathContainer`, I think adding it to this PR is more fitting since its a container implementation. #7 is specifically related to created hierarchical structures.

Reply:
Fair point.